### ISSUE TITLE
Rule for detecting generator functions without yield statements

### ISSUE THREAD

**Issue by moll** [eslint/eslint#2822]
Hey,

It'd be useful to get a warning when a generator function is written without a `yield` statement inside it.
Would match JSHint's `noyield`: http://jshint.com/docs/options/#noyield.

Thanks!

**Comment by nzakas** [comment]
This sounds like a good idea and I'd like to hear more. Please provide the information asked for in our new rule submission guide: http://eslint.org/docs/developer-guide/contributing#new-rules

**Comment by moll** [comment]
This rule would catch unnecessary generator functions where a plain, synchronous function would do. Helpful in server-side contexts where generators are used for writing sequential code.

### Code examples

Invalid:
```javascript
function*() {
  return 42
}
```

Valid:
```javascript
function*() {
  yield 13  
  return 42
}
```

**Comment by nzakas** [comment]
Thanks. Are you interested in working on this?

**Comment by moll** [comment]
I would love to, but I'm probably not too likely to get to it in the nearer future. I've got so many open pull requests, issues and todos on my own OSS projects that I'm a little overwhelmed. :innocent: 

Hopefully someone better abled comes along to make this rule happen.

**Comment by mysticatea** [comment]
Can I try to make this?
I will name this rule `no-extra-generator`, which is similar to `no-extra-bind`.

**Comment by ilyavolodin** [comment]
I don't think that's a descriptive name for this rule. I would ideally call it `no-generator-without-yield`, but that's too long. Maybe something like `require-iterator`, but that doesn't follow our naming convention...

**Comment by mathieumg** [comment]
> I would ideally call it no-generator-without-yield, but that's too long.

I think it's a good descriptive name, and its name is as long as http://eslint.org/docs/rules/no-shadow-restricted-names

**Comment by nzakas** [comment]
Maybe require-yield?

The no- prefix we typically only use when we want to prevent you from doing something. In this case, I'd argue we are trying to get you to do something.

**Comment by ilyavolodin** [comment]
Hmm.. I think I agree. `require-yield` is good name from my perspective, I would go with that.

**Comment by mysticatea** [comment]
Thank you for naming advice!
